SABT_MainUnit/MD4.c

```diff
@@ -25,293 +36,325 @@
-      // nothing was entered so we repeat the word
-      if (last_cell == 0)
-      {
-        md4_current_state = MD4_STATE_EVALUATE_GAME;
-      } else if (md4_valid_letter(last_cell))
-      { // set entered_letter in valid_letter(), but return true or false
-        char buff[7];
-        sprintf(buff, "%c", entered_letter);
-        play_mp3(LANG_FILESET,buff);
-        md4_current_state = MD4_STATE_CHECK_MATCH;
-      } else
-      {
-        play_mp3(LANG_FILESET,MP3_INVALID);
-        num_mistakes++;
-        md4_current_state = MD4_STATE_EVALUATE_GAME;
-      }
-      break;
 
-    case MD4_STATE_CHECK_MATCH:
-      // place_letter() returns true if the letter guessed is found in
-      // the word, false otherwise. If true, it will place the entered_letter
-      // into input_word.
-      if (place_letter())
-      {
-        play_mp3(LANG_FILESET,MP3_YES);
-      } else
-      {
-        play_mp3(LANG_FILESET, MP3_NO);
-        num_mistakes++;
-      }
-
-      md4_current_state = MD4_STATE_EVALUATE_GAME;
-      break;
+void md4_reset(void) {
+    PRINTF("*** MD4 - one player hangman ***\n\r");
+    
+    // Global variables
+    set_mode_globals(&script_english, LANG_FILESET, MODE_FILESET);
+    
+    // State variables
+    md_next_state = STATE_INTRO;
+    md_last_dot = NO_DOTS;
+    md_input_ready = false;
+    md_input_valid = false;
+    md_incorrect_tries = 0;
+}
 
-    case MD4_STATE_EVALUATE_GAME:
-      if (!strncmp(input_word, current_word, strlen(current_word)))
-      {
-        game_status = 1;
-        play_mp3(MODE_FILESET,MP3_YOU_WIN);  // "you have guessed the word!"
-      } else if (num_mistakes == 7)
-      {
-        game_status = 1;
-        play_mp3(MODE_FILESET,MP3_YOU_LOSE); // "you have made 7 mistakes the word you missed was"
-      }
+void md4_main(void) {
+	int num_words;	
+    switch (md_next_state) {
+        case STATE_INTRO:
+			shuffle(dict.num_words, dict.index_array);
+			dict.index = 0;
+            play_mp3(MODE_FILESET, MP3_INTRO);
+            md_next_state = STATE_GENQUES;
+            break;
+            
+        case STATE_GENQUES:
+			
+            num_words = dict.num_words;
+            if (dict.index >= num_words - 1){
+                shuffle(num_words, dict.index_array);
+                dict.index = 0;
+                }
+            chosen_word = dict.words[dict.index_array[dict.index]];
+            dict.index++;
+            
+            char buf[10];
+            sprintf(buf, "word:%s\r\n", chosen_word);
+            PRINTF(buf);
+			md_next_state = STATE_LVLSEL;
+            
+            break;
+            
+        case STATE_LVLSEL:
+			num_mistakes = 0;
+            init_char_arr(mistake_pool, MAX_INCORRECT_GUESS);
+            init_char_arr(input_word, MAX_WORD_LEN);
+            md_last_dot = create_dialog(MP3_LEVEL,
+                                        DOT_1 | DOT_2 | DOT_3 );
+            switch (md_last_dot) {
+					case NO_DOTS:
+						break;

```

Comment:
When trying to press Enter for no hints, the SABT said "invalid pattern".  Maybe some change around here?